{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "95c5ce0e-ff86-439b-b1a5-6dcbd709251e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-11-22 18:24:54,553 - main_logger - INFO - Este es un mensaje de información.\n",
      "2024-11-22 18:24:54,555 - main_logger - WARNING - Este es un mensaje de advertencia.\n",
      "2024-11-22 18:24:54,556 - main_logger - ERROR - Este es un mensaje de error.\n",
      "2024-11-22 18:24:54,557 - main_logger - CRITICAL - Este es un mensaje crítico.\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/python/3.12.1/lib/python3.12/logging/__init__.py\", line 1160, in emit\n",
      "    msg = self.format(record)\n",
      "          ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/python/3.12.1/lib/python3.12/logging/__init__.py\", line 999, in format\n",
      "    return fmt.format(record)\n",
      "           ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_704/184533936.py\", line 44, in format\n",
      "    return f'{record.asctime},{record.name},{record.levelname},{record.message}'\n",
      "              ^^^^^^^^^^^^^^\n",
      "AttributeError: 'LogRecord' object has no attribute 'asctime'\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/usr/local/python/3.12.1/lib/python3.12/asyncio/base_events.py\", line 638, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/usr/local/python/3.12.1/lib/python3.12/asyncio/base_events.py\", line 1971, in _run_once\n",
      "    handle._run()\n",
      "  File \"/usr/local/python/3.12.1/lib/python3.12/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/codespace/.local/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_704/184533936.py\", line 54, in <module>\n",
      "    csv_logger.info(\"Mensaje de prueba en formato CSV.\")\n",
      "Message: 'Mensaje de prueba en formato CSV.'\n",
      "Arguments: ()\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "from logging import Formatter, FileHandler\n",
    "\n",
    "# Configuración del logger principal\n",
    "logger = logging.getLogger('main_logger')\n",
    "logger.setLevel(logging.DEBUG)\n",
    "\n",
    "# Handler para logs de error\n",
    "error_handler = logging.FileHandler('error_logs.log')\n",
    "error_handler.setLevel(logging.ERROR)\n",
    "\n",
    "# Handler para logs de info\n",
    "info_handler = logging.FileHandler('info_logs.log')\n",
    "info_handler.setLevel(logging.INFO)\n",
    "\n",
    "# Handler para mostrar logs por pantalla\n",
    "console_handler = logging.StreamHandler()\n",
    "console_handler.setLevel(logging.INFO)\n",
    "\n",
    "# Formateadores para los logs\n",
    "formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')\n",
    "error_handler.setFormatter(formatter)\n",
    "info_handler.setFormatter(formatter)\n",
    "console_handler.setFormatter(formatter)\n",
    "\n",
    "# Agregar handlers al logger principal\n",
    "logger.addHandler(error_handler)\n",
    "logger.addHandler(info_handler)\n",
    "logger.addHandler(console_handler)\n",
    "\n",
    "# Agregar mensajes de advertencia\n",
    "logger.info(\"Este es un mensaje de información.\")\n",
    "logger.warning(\"Este es un mensaje de advertencia.\")\n",
    "logger.error(\"Este es un mensaje de error.\")\n",
    "logger.critical(\"Este es un mensaje crítico.\")\n",
    "\n",
    "# Configuración del logger especial con formato CSV\n",
    "csv_logger = logging.getLogger('csv_logger')\n",
    "csv_logger.setLevel(logging.INFO)\n",
    "\n",
    "# Handler con formato CSV\n",
    "class CSVFormatter(Formatter):\n",
    "    def format(self, record):\n",
    "        return f'{record.asctime},{record.name},{record.levelname},{record.message}'\n",
    "\n",
    "csv_handler = FileHandler('csv_logs.csv', mode='w')\n",
    "csv_handler.setLevel(logging.INFO)\n",
    "csv_handler.setFormatter(CSVFormatter())\n",
    "\n",
    "# Agregar handler al logger especial\n",
    "csv_logger.addHandler(csv_handler)\n",
    "\n",
    "# Generar un mensaje de prueba para el logger CSV\n",
    "csv_logger.info(\"Mensaje de prueba en formato CSV.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4599e315-2aaa-477e-b3e8-8143a373a9ec",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
